primitives/api/proc-macro/src/impl_runtime_apis.rs

```diff
@@ -248,11 +248,13 @@ fn generate_runtime_api_base_structures() -> Result<TokenStream> {
 				) -> R where Self: Sized {
 					self.start_transaction();
 
-					*std::cell::RefCell::borrow_mut(&self.commit_on_success) = false;
+					let old_value = std::cell::RefCell::replace(&self.in_transaction, true);
 					let res = call(self);
-					*std::cell::RefCell::borrow_mut(&self.commit_on_success) = true;
+					*std::cell::RefCell::borrow_mut(&self.in_transaction) = old_value;
```

Comment:
This might not matter in practice (?), but `commit_or_rollback_transaction` can panic, and now we're going to support nested transactions, which means that `call` can panic, which means that if it does then the `in_transaction` won't get restored to the old value.

Reply:
Yeah I know, but the type should not be unwind safe (which is the case). However, I added a test to ensure this.

Reply:
Hmmm... to be honest, I'm not sure how useful that is considering how borderline useless the `UnwindSafe` trait is. (Since e.g. even `&mut T` ends up being not `UnwindSafe`, which makes it extremely false positive-y, which makes the use of `AssertUnwindSafe` very common.)

Maybe instead of `in_transaction: bool` we could have e.g. `transaction_depth: usize` and increment/decrement it when we enter/leave a transaction? I think that'd be a little nicer. (And in case there's any funny business going to it'd be easier to debug since you'd have a stale non-zero counter.)

But it's up to you.